lib/charms/identity_platform_login_ui_operator/v0/login_ui_endpoints.py

```diff
@@ -0,0 +1,171 @@
+
+
+class LoginUIEndpointsProviderEvents(ObjectEvents):
+    """Event descriptor for events raised by `LoginUIEndpointsProvider`."""
+
+    ready = EventSource(LoginUIEndpointsRelationReadyEvent)
+
+
+class LoginUIEndpointsProvider(Object):
+    """Provider side of the endpoint-info relation."""
+
+    on = LoginUIEndpointsProviderEvents()
+
+    def __init__(self, charm: CharmBase, relation_name: str = RELATION_NAME):
+        super().__init__(charm, relation_name)
+
+        self._charm = charm
+        self._relation_name = relation_name
+
+        events = self._charm.on[relation_name]
+        self.framework.observe(
+            events.relation_created, self._on_provider_endpoints_relation_created
+        )
+
+    def _on_provider_endpoints_relation_created(self, event: RelationCreatedEvent) -> None:
+        self.on.ready.emit()
+
+    def send_endpoints_relation_data(self, endpoint: str) -> None:
+        """Updates relation with endpoint info."""
+        if not self._charm.unit.is_leader():
+            return
+
+        relations = self.model.relations[self._relation_name]
+        for relation in relations:
+            relation.data[self._charm.app].update(
+                {
+                    "consent_url": f"{endpoint}/consent",
+                    "error_url": f"{endpoint}/error",
+                    "index_url": f"{endpoint}/index",
+                    "login_url": f"{endpoint}/login",
+                    "oidc_error_url": f"{endpoint}/oidc_error",
+                    "registration_url": f"{endpoint}/registration",
+                    "default_url": endpoint,
+                }
+            )
+
+
+class LoginUIEndpointsRelationError(Exception):
+    """Base class for the relation exceptions."""
+
+    pass
+
+
+class LoginUIEndpointsRelationMissingError(LoginUIEndpointsRelationError):
+    """Raised when the relation is missing."""
+
+    def __init__(self) -> None:
+        self.message = "Missing ui-endpoint-info relation with Identity Platform Login UI"
+        super().__init__(self.message)
+
+
+class LoginUIEndpointsRelationDataMissingError(LoginUIEndpointsRelationError):
+    """Raised when information is missing from the relation."""
+
+    def __init__(self, message: str) -> None:
+        self.message = message
+        super().__init__(self.message)
+
+
+class LoginUIEndpointsRelationUnavailableError(LoginUIEndpointsRelationError):
+    """Raised when Login UI cannot be accessed."""
+
+    def __init__(self, message: str) -> None:
+        self.message = message
+        super().__init__(self.message)
+
+
+class LoginUIEndpointsRequirer(Object):
+    """Requirer side of the ui-endpoint-info relation."""
+
+    def __init__(self, charm: CharmBase, relation_name: str = RELATION_NAME):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self._relation_name = relation_name
+
+    def get_login_ui_endpoints(self) -> Optional[Dict]:
```

Comment:
```suggestion
    def get_login_ui_endpoints(self) -> Dict:
```

Reply:
`
        if any(not data.get(k := key) for key in RELATION_KEYS):
            raise LoginUIEndpointsRelationDataMissingError(
                f"Missing endpoint {k} in ui-endpoint-info relation data"
            )
`
I'm 90% sure this line returns the function with None if there's a missing or empty value for a key.

Reply:
Raising an exception is different from returning `None`. You can't express exceptions with Python type hints (https://peps.python.org/pep-0484/#exceptions)

Reply:
OK, I changed the return directive